Add progress tracking to ffmpeg operations

Kenshin9977 opened the pull request:
## Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [X] Improvement
- [ ] New extractor
- [X] New feature

---

### Description

Add real-time progress tracking to ffmpeg operations, both for downloading (`FFmpegFD`) and post-processing (`FFmpegPostProcessor`).

#### How it works

A new `FFmpegProgressTracker` class manages ffmpeg subprocess I/O using a **Queue + Thread** pattern:

1. ffmpeg is launched with `-progress pipe:1`, which outputs machine-readable progress data to stdout
2. Two daemon threads read stdout/stderr line-by-line into thread-safe queues
3. The main thread polls at 50ms intervals, draining the queues and parsing progress blocks
4. Parsed progress (ETA, speed, bytes outputted) is reported through the existing progress hook chain

#### Changes

**`yt_dlp/postprocessor/ffmpeg.py`**
- New `FFmpegProgressTracker` class (~250 lines) handling subprocess lifecycle, progress parsing, ETA/bitrate/filesize computation
- `real_run_ffmpeg` now creates a tracker and passes `-progress pipe:1` when `information` is provided
- Progress hooks translate ffmpeg-internal status (`ffmpeg_running` → `processing`, `outputted` → `processed_bytes`)

**`yt_dlp/downloader/external.py`**
- `FFmpegFD` now uses `FFmpegProgressTracker` instead of raw `Popen`
- Always appends `-progress pipe:1` since `info_dict` is always available in the downloader context
- Progress hooks translate status (`ffmpeg_running` → `downloading`, `outputted` → `downloaded_bytes`)

**`yt_dlp/postprocessor/common.py`**
- `report_progress` handles the `processing` status with percent, speed, ETA display
- Gracefully handles minimal status dicts from `PostProcessorMetaClass.run_wrapper`

**`test/test_ffmpeg_progress.py`** — 82 unit tests covering:
- Progress regex pattern matching (single/multi stream, audio-only, N/A values)
- Time string parsing (HMS, ms/us units, fractional seconds)
- Duration computation with seek args (`-ss`, `-to`, `-t`, `-sseof`, `=` form)
- Bitrate parsing, ETA computation, filesize estimation
- Queue draining, stream accumulation, progress reporting

Closes https://github.com/yt-dlp/yt-dlp/issues/1720, https://github.com/yt-dlp/yt-dlp/issues/2826, Closes #1317
Replaces https://github.com/yt-dlp/yt-dlp/pull/1133, https://github.com/yt-dlp/yt-dlp/pull/1947

Kenshin9977 committed:
Add ffmpeg progress tracking to FFmpegFD

Add ffmpeg progress tracking to FFmpegPostProcessor

Apply changes from the code review

Fix a bug where the subprocess didn't capture any output thus an empty stdout and stderr were sent back

Add missing hooks

Revert "Add missing hooks"

This reverts commit a359c5ea10bb35b965e80801e736f43cdbcf3294.

Add support of -ss=132 timestamp format

Infer filename from ffmpeg args instead of info_dic

Remove redundant parenthesis and switch from to_stodout to to_screen

Add info kwarg with multiple files and ffmpeg to track progress

Moved format progress function to util

Moved format progress function to util

Add progress tracking to postprocessing operations

Fix typing error

Handle self._downloader is None at __init__

Move format progress functions to utils

Move format progress functions to utils

Handle case where ydl passed is None

Handle case where ydl passed is None

Handle case where _multiline isn't initialized

Handle case where _multiline isn't initialized

Fix streams incorrectly returned

Fix case where ydl is nested in the downloader

Add progress_hook attribute

Fix bug after merge

Fix import bugs after merge

Catch up with upstream

Fix merge errors #1

Adapt tests and implementatation for ffmpeg progress tracking args

pukkandan commented on yt_dlp/downloader/external.py:
```diff
@@ -481,13 +482,16 @@ def _call_downloader(self, tmpfilename, info_dict):
         args.append(encodeFilename(ffpp._ffmpeg_filename_argument(tmpfilename), True))
+        args.extend(['-progress', 'pipe:1', '-stats_period', '0.1'])
         self._debug_cmd(args)
 
-        proc = Popen(args, stdin=subprocess.PIPE, env=env)
+        proc = Popen(args, env=env, universal_newlines=True, encoding='utf8', stdin=subprocess.PIPE, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
         if url in ('-', 'pipe:'):
             self.on_process_started(proc, proc.stdin)
         try:
-            retval = proc.wait()
+            ffmpeg_progress_tracker = FFmpegProgressTracker(info_dict, args, proc, self._hook_progress, 'downloading', 'downloaded_bytes')
```
Since all the details needed for starting the process is being passed to FFmpegProgressTracker, why not let it handle that as well. So something like this:

```py
        ffmpeg_progress_tracker = FFmpegProgressTracker(...)
        if url in ('-', 'pipe:'):
            proc = ffmpeg_progress_tracker.process
            self.on_process_started(proc, proc.stdin)
```

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -315,26 +316,38 @@ def make_args(file, args, name, number):
-            cmd += itertools.chain.from_iterable(
-                make_args(path, list(opts), arg_type, i + 1)
-                for i, (path, opts) in enumerate(path_opts) if path)
+            for i, (path, opts) in enumerate(path_opts):
+                if path:
+                    cmd += make_args(path, list(opts), arg_type, i + 1)
```
why?

Kenshin9977 replied:
I felt like the original code was overly complicated with no good reason. This expression is way easier to read, takes the same amout of lines and doesn't increase runtime

pukkandan replied:
I disagree. The current implementation is fine. Besides, this is not a re-factoring PR. Do not make unrelated changes

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -315,26 +316,38 @@ def make_args(file, args, name, number):
+        # Needed to track ffmpeg progress
+        cmd.extend(['-progress', 'pipe:1', '-stats_period', '0.1'])
+
+        if type(input_path_opts) == list:
+            file_path, _ = input_path_opts[0]
+        elif type(input_path_opts) == tuple:
+            file_path, _ = input_path_opts
+        else:
+            raise ValueError(f"input_path_opts should either be a tuple or a list but it is a '{type(input_path_opts)}'")
```
this is wrong. If you are looking for the first input file, it should be:
```suggestion
        file_path = input_path_opts[0][0]
```
But I don't understand what you are trying to do here. `info_dict` passed to the hook should be the actual info_dict, not this dummy dict

Kenshin9977 replied:
info_dict doesn't exist in this context and it doesn't exists everywhere where this function is called

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -315,26 +316,38 @@ def make_args(file, args, name, number):
-        stdout, stderr = p.communicate_or_kill()
-        if p.returncode not in variadic(expected_retcodes):
-            stderr = stderr.decode('utf-8', 'replace').strip()
+        file_infos = self.get_metadata_object(file_path)
+        filename = os.path.basename(file_path)
+        info_dict = {'_filename': filename, 'duration': file_infos.get('duration'), 'filesize': file_infos.get('filesize')}
```
see above

Lesmiscore replied:
I don't see point of getting `duration` using ffprobe though
Duration from extractor can be obtained by `info_dict.get('duration')`, also if it's missing why don't we leave it absent?

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1173,3 +1186,160 @@ def run(self, info):
+        self.ffmpeg_args = ffmpeg_args
+        self.ffmpeg_proc = ffmpeg_proc
+        self.hook_progress = hook_progress
+        self.running_status_value = running_status_value
+        self.action_bytes_key = action_bytes_key
```
this seems like a gimick. Better to pass a local hook function here which translates the keys to be suitable for the actual `hook_progress`. We can handle more cases in future like that

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -5,9 +5,11 @@
 import time
 import re
 import json
+from typing import Match, Callable, Tuple, List
```
pls remove

Kenshin9977 replied:
I shouldn't add type hints?

pukkandan replied:
Yes

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1173,3 +1186,160 @@ def run(self, info):
+
+class FFmpegProgressTracker:
+
+    def __init__(self, info_dict: dict, ffmpeg_args: List, ffmpeg_proc: Popen, hook_progress: Callable, running_status_value: str, action_bytes_key: str):
```
why not actually needed rn, pass the `YoutubeDL` instance to the initializer as `self.ydl`. Otherwise we'll be stuck if we need to access any of it's functions/params in future

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1173,3 +1186,160 @@ def run(self, info):
+        total_filesize = self.compute_total_filesize(duration_to_track, duration)
+        started = time.time()
+        status = {
+            'filename': self.info_dict.get('_filename'),
```
technically, this should be the output filename

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1173,3 +1186,160 @@ def run(self, info):
+            'elapsed': time.time() - started
+        }
+        # Parse ffmpeg_output
+        progress_pattern = re.compile(
+            r'(frame=\s*(?P<frame>\S+)\nfps=\s*(?P<fps>\S+)\nstream_0_0_q=\s*(?P<stream_0_0_q>\S+)\n)?bitrate=\s*(?P<bitrate>\S+)\ntotal_size=\s*(?P<total_size>\S+)\nout_time_us=\s*(?P<out_time_us>\S+)\nout_time_ms=\s*(?P<out_time_ms>\S+)\nout_time=\s*(?P<out_time>\S+)\ndup_frames=\s*(?P<dup_frames>\S+)\ndrop_frames=\s*(?P<drop_frames>\S+)\nspeed=\s*(?P<speed>\S+)\nprogress=\s*(?P<progress>\S+)')
```
make a class var. And format better using `r'''(?x)...'''`

Lesmiscore replied:
I'm wondering if there's more and more streams in output file.
ref. https://ffmpeg.org/pipermail/ffmpeg-user/2016-July/032897.html

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1173,3 +1186,160 @@ def run(self, info):
+                return ffmpeg_stdout_buffer, ffmpeg_stderr_buffer, 0
+
+        status.update({
+            'status': 'finished',
```
`finished` hook should invoked only when the entire PP has run. A PP may have multiple ffmpeg runs. But it's fine to return finished here and we can handle it in the translation layer imo

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -315,26 +316,38 @@ def make_args(file, args, name, number):
-        stdout, stderr = p.communicate_or_kill()
-        if p.returncode not in variadic(expected_retcodes):
-            stderr = stderr.decode('utf-8', 'replace').strip()
+        file_infos = self.get_metadata_object(file_path)
```
ffprobe should not become a dependency for operations that actually only need ffmpeg

Lesmiscore commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1173,3 +1186,160 @@ def run(self, info):
+        ffmpeg_stderr_buffer = ''
+        while retval is None:
+            ffmpeg_stdout = self.ffmpeg_proc.stdout.readline() if self.ffmpeg_proc.stdout else ''
+            ffmpeg_stderr_buffer = self.ffmpeg_proc.stderr.readline() if self.ffmpeg_proc.stderr else ''
```
using `readline()` to stderr from ffmpeg will block until ffmpeg finishes, since it doesn't emit break lines (LF in Linux/macOS, CRLF in Windows) except banner and end of process.

Kenshin9977 replied:
By default ffmpeg outputs on stderr. When I test my PR, it outputs both on stderr and stdout during the whole process andn ot only when it finishes

Lesmiscore replied:
~~since it'll be adding `-progress pipe:1`, status will go to stdout, but visual progress isn't and has no newlines~~
honestly I'm not confident 😅

Kenshin9977 replied:
It has newlines. Don't worry, I tested it works before submitting this PR :D

Lesmiscore replied:
https://github.com/ytdl-patched/ytdl-patched/blob/feba3d336892845cef553436818170554cfa5228/yt_dlp/postprocessor/ffmpeg.py#L355-L356
Looks like I'm mistaken why it was stuck

Lesmiscore commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1173,3 +1186,160 @@ def run(self, info):
+            ffmpeg_prog_infos = re.match(progress_pattern, ffmpeg_stdout_buffer)
+            if not ffmpeg_prog_infos:
+                continue
+            sys.stdout.write(ffmpeg_stdout_buffer)
```
wouldn't it break `-o -`?

Kenshin9977 replied:
I'm sorry I'll need more context because I don't understand your comment

Lesmiscore replied:
~~yt-dlp has an option to write downloaded data to stdout (`-o -`) blablabla~~
Ah when it has `-o -` it should work ig

Lesmiscore reviewed (commented):
[Writing my experience while modifying your PR.](https://github.com/ytdl-patched/ytdl-patched/blob/ytdlp/yt_dlp/postprocessor/_attachments.py)

pukkandan commented:
Some of the newer commits are seriously overreaching beyond the scope of this PR. Please revert them

pukkandan commented on yt_dlp/downloader/external.py:
```diff
@@ -481,13 +482,15 @@ def _call_downloader(self, tmpfilename, info_dict):
-        proc = Popen(args, stdin=subprocess.PIPE, env=env)
+        self._debug_cmd(args)
+        ffmpeg_progress_tracker = FFmpegProgressTracker(info_dict, args, self._ffmpeg_hook)
+        proc = ffmpeg_progress_tracker.ffmpeg_proc
```
this is wrong. See below comment

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1169,3 +1178,226 @@ def run(self, info):
+            'outputted': 0
+        }
+
+        self.ffmpeg_proc = Popen(
+            self._ffmpeg_args,
+            env=environ.copy(),
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            universal_newlines=True,
+            encoding='utf8',
+        )
```
You already created a process at init

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1169,3 +1178,226 @@ def run(self, info):
+            return self._track_ffmpeg_progress()
+        return self._run_ffmpeg_without_progress_tracking()
+
+    def _run_ffmpeg_without_progress_tracking(self):
+        """Simply run ffmpeg and only care about the last stderr, stdout and the retcode"""
+        stdout, stderr = self.ffmpeg_proc.communicate_or_kill()
+        retcode = self.ffmpeg_proc.returncode
+        return stdout, stderr, retcode
```
Why is this needed?

Kenshin9977 replied:
Thought of it as a way to slightly optimize the ffmpeg run. This way it won't uselessly parse the output thus saving a small bit of resources

Kenshin9977 replied:
Everywhere there is a call to `real_run_ffmpeg` and where info_dict doesn't exist. From `def concat_files` for instance. That's why I added `information` as an argument to `real_run_ffmpeg` with its default value to `None` as suggested by @Lesmiscore
If that's not a valid implementation of if I understood something wrong I will obviously change it but I need a hint in order to go in the right direction

Lesmiscore replied:
suggestion: do `info_dict = info_dict or {}` somewhere, and hooks should be called

since not all calls (afaik) can get info_dict, I think that's okay to skip hooks though

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1169,3 +1178,226 @@ def run(self, info):
+        })
+        self._hook_progress(self._status, self._info_dict)
+        self._stderr_buffer = re.sub(r'=\s+', '=', self._stderr_buffer)
+        print(self._stdout_buffer, file=sys.stdout, end='')
+        print(self._stderr_buffer, file=sys.stderr)
+        self._stdout_buffer = ''
+        self._stderr_buffer = ''
```
?

Kenshin9977 replied:
Don't we want to print ffmpeg output?

pukkandan replied:
It is important that we at least don't print the progress lines we just parsed.

Ideally, we also don't want to print anything in stdout unless in verbose mode. It would be nice if warnings can obey `--no-warnings`, but not sure if that is feasible to implement. Errors should always be printed

pukkandan replied:
Also, use to_screen, to_stderr etc, not print

pukkandan commented:
```
❯ yt-dlp --downloader ffmpeg -x test:youtube_4 --force-over -v
[debug] Command-line config: ['--downloader', 'ffmpeg', '-x', 'test:youtube_4', '--force-over', '-v']
[debug] Encodings: locale cp65001, fs utf-8, out utf-8, err utf-8, pref cp65001
[debug] yt-dlp version 2022.02.04 [c1653e9ef] (source)
[debug] Lazy loading extractors is disabled
[debug] Plugins: ['SamplePluginIE', 'SamplePluginPP']
[debug] Git HEAD: a9eed42f8
[debug] Python version 3.10.4 (CPython 64bit) - Windows-10-10.0.22000-SP0
[debug] exe versions: ffmpeg N-105465-gbedb4bac1c-20220206 (setts,fdk), ffprobe N-105465-gbedb4bac1c-20220206, phantomjs 2.1.1
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
[debug] [TestURL] Extracting URL: test:youtube_4
[TestURL] Test URL: https://www.youtube.com/watch?v=IB3lcPjvWLA
[debug] [youtube] Extracting URL: https://www.youtube.com/watch?v=IB3lcPjvWLA
[youtube] IB3lcPjvWLA: Downloading webpage
[youtube] IB3lcPjvWLA: Downloading android player API JSON
[debug] Sort order given by extractor: quality, res, fps, hdr:12, source, codec:vp9.2, lang, proto
[debug] Formats sorted by: hasvid, ie_pref, quality, res, fps, hdr:12(7), source, vcodec:vp9.2(10), acodec, lang, proto, filesize, fs_approx, tbr, vbr, abr, asr, vext, aext, hasaud, id
[info] IB3lcPjvWLA: Downloading 1 format(s): 251
Deleting existing file Afrojack, Spree Wilson - The Spark (Official Music Video) ft. Spree Wilson [IB3lcPjvWLA].webm
[debug] Invoking downloader on "https://rr3---sn-gwpa-jv3z.googlevideo.com/videoplayback?expire=1649458283&ei=C2hQYq_VMOWI9fwPo8usoAs&ip=49.37.227.151&id=o-AOGmcyslqPHAg9393oMjWNZxeR4XDOekMJGjAVMWY9H_&itag=251&source=youtube&requiressl=yes&mh=qo&mm=31%2C29&mn=sn-gwpa-jv3z%2Csn-gwpa-jwce&ms=au%2Crdu&mv=m&mvi=3&pl=20&gcr=in&initcwndbps=200000&vprv=1&mime=audio%2Fwebm&gir=yes&clen=4099816&dur=243.561&lmt=1540285801235469&mt=1649436303&fvip=3&keepalive=yes&fexp=24001373%2C24007246&c=ANDROID&txp=5511222&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cgcr%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgeRE670-S1-5VxFGq8in-0wKah3ykZrqSNTdwMCvTdqwCIAY7-Fcz1dSWAkt_RWzTLRux_SfbOG1Qk0x8kS28z9uX&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIgWdDRfENdvxBOgMCoUlDYyAKYypOlemnVcVsmaQZQmhgCIQC4mszPG_ibErJr2eqMaC2wAb3IDfwGbf6yqpC33Esxzg%3D%3D"
[download] Destination: Afrojack, Spree Wilson - The Spark (Official Music Video) ft. Spree Wilson [IB3lcPjvWLA].webm
[debug] ffmpeg command line: ffmpeg -y -loglevel verbose -headers "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.81 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Encoding: gzip, deflate
Accept-Language: en-us,en;q=0.5
Sec-Fetch-Mode: navigate
" -i "https://rr3---sn-gwpa-jv3z.googlevideo.com/videoplayback?expire=1649458283&ei=C2hQYq_VMOWI9fwPo8usoAs&ip=49.37.227.151&id=o-AOGmcyslqPHAg9393oMjWNZxeR4XDOekMJGjAVMWY9H_&itag=251&source=youtube&requiressl=yes&mh=qo&mm=31%2C29&mn=sn-gwpa-jv3z%2Csn-gwpa-jwce&ms=au%2Crdu&mv=m&mvi=3&pl=20&gcr=in&initcwndbps=200000&vprv=1&mime=audio%2Fwebm&gir=yes&clen=4099816&dur=243.561&lmt=1540285801235469&mt=1649436303&fvip=3&keepalive=yes&fexp=24001373%2C24007246&c=ANDROID&txp=5511222&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cgcr%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgeRE670-S1-5VxFGq8in-0wKah3ykZrqSNTdwMCvTdqwCIAY7-Fcz1dSWAkt_RWzTLRux_SfbOG1Qk0x8kS28z9uX&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIgWdDRfENdvxBOgMCoUlDYyAKYypOlemnVcVsmaQZQmhgCIQC4mszPG_ibErJr2eqMaC2wAb3IDfwGbf6yqpC33Esxzg%3D%3D" -c copy -f webm "file:Afrojack, Spree Wilson - The Spark (Official Music Video) ft. Spree Wilson [IB3lcPjvWLA].webm.part"
ffmpeg version N-105465-gbedb4bac1c-20220206 Copyright (c) 2000-2022 the FFmpeg developers
  built with gcc 11.2.0 (crosstool-NG 1.24.0.498_5075e1f)
  configuration: --prefix=/ffbuild/prefix --pkg-config-flags=--static --pkg-config=pkg-config --cross-prefix=x86_64-w64-mingw32- --arch=x86_64 --target-os=mingw32 --enable-nonfree --enable-gpl --enable-version3 --disable-debug --disable-w32threads --enable-pthreads --enable-iconv --enable-libxml2 --enable-zlib --enable-libfreetype --enable-libfribidi --enable-gmp --enable-lzma --enable-fontconfig --enable-libvorbis --enable-opencl --disable-libpulse --enable-libvmaf --disable-libxcb --disable-xlib --enable-amf --enable-libaom --enable-avisynth --enable-libdav1d --enable-libdavs2 --enable-libfdk-aac --enable-ffnvcodec --enable-cuda-llvm --enable-frei0r --enable-libgme --enable-libass --enable-libbluray --enable-libmp3lame --enable-libopus --enable-librist --enable-libtheora --enable-libvpx --enable-libwebp --enable-lv2 --enable-libmfx --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenh264 --enable-libopenjpeg --enable-libopenmpt --enable-librav1e --enable-librubberband --enable-schannel --enable-sdl2 --enable-libsoxr --enable-libsrt --enable-libsvtav1 --enable-libtwolame --enable-libuavs3d --disable-libdrm --disable-vaapi --enable-libvidstab --enable-vulkan --enable-libshaderc --enable-libplacebo --enable-libx264 --enable-libx265 --enable-libxavs2 --enable-libxvid --enable-libzimg --enable-libzvbi --extra-cflags=-DLIBTWOLAME_STATIC --extra-cxxflags= --extra-ldflags=-pthread --extra-ldexeflags= --extra-libs=-lgomp --extra-version=20220206
  libavutil      57. 19.100 / 57. 19.100
  libavcodec     59. 20.100 / 59. 20.100
[download]   0.0% of 3.91MiB at Unknown speed ETA Unknown  libavformat    59. 17.101 / 59. 17.101
  libavdevice    59.  5.100 / 59.  5.100
  libavfilter     8. 26.101 /  8. 26.101
  libswscale      6.  5.100 /  6.  5.100
  libswresample   4.  4.100 /  4.  4.100
  libpostproc    56.  4.100 / 56.  4.100
[download]   0.0% of 3.91MiB at Unknown speed ETA Unknown[tcp @ 000002996651ce80] Starting connection attempt to 49.44.85.46 port 443
[download]   0.0% of 3.91MiB at Unknown speed ETA UnknownSuccessfully connected to 49.44.85.46 port 443
[download]   0.0% of 3.91MiB at Unknown speed ETA UnknownInput #0, matroska,webm, from 'https://rr3---sn-gwpa-jv3z.googlevideo.com/videoplayback?expire=1649458283&ei=C2hQYq_VMOWI9fwPo8usoAs&ip=49.37.227.151&id=o-AOGmcyslqPHAg9393oMjWNZxeR4XDOekMJGjAVMWY9H_&itag=251&source=youtube&requiressl=yes&mh=qo&mm=31%2C29&mn=sn-gwpa-jv3z%2Csn-gwpa-jwce&ms=au%2Crdu&mv=m&mvi=3&pl=20&gcr=in&initcwndbps=200000&vprv=1&mime=audio%2Fwebm&gir=yes&clen=4099816&dur=243.561&lmt=1540285801235469&mt=1649436303&fvip=3&keepalive=yes&fexp=24001373%2C24007246&c=ANDROID&txp=5511222&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cgcr%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgeRE670-S1-5VxFGq8in-0wKah3ykZrqSNTdwMCvTdqwCIAY7-Fcz1dSWAkt_RWzTLRux_SfbOG1Qk0x8kS28z9uX&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIgWdDRfENdvxBOgMCoUlDYyAKYypOlemnVcVsmaQZQmhgCIQC4mszPG_ibErJr2eqMaC2wAb3IDfwGbf6yqpC33Esxzg%3D%3D':
  Metadata:
    encoder         : google/video-file
  Duration: 00:04:03.56, start: -0.007000, bitrate: 134 kb/s
  Stream #0:0(eng): Audio: opus, 48000 Hz, stereo, fltp, delay 312 (default)
Output #0, webm, to 'file:Afrojack, Spree Wilson - The Spark (Official Music Video) ft. Spree Wilson [IB3lcPjvWLA].webm.part':
  Metadata:
    encoder         : Lavf59.17.101
  Stream #0:0(eng): Audio: opus, 48000 Hz, stereo, fltp, delay 312 (default)
Stream mapping:
  Stream #0:0 -> #0:0 (copy)
Press [q] to stop, [?] for help
[download]   0.0% of 3.91MiB at    0.00B/s ETA 00:00bitrate=N/A
total_size=475
out_time_us=0
out_time_ms=0
out_time=00:00:00.000000
dup_frames=0
drop_frames=0
speed=   0x
progress=continue
ffmpeg version N-105465-gbedb4bac1c-20220206 Copyright (c) 2000-2022 the FFmpeg developers  built with gcc 11.2.0 (crosstool-NG 1.24.0.498_5075e1f)  configuration: --prefix=/ffbuild/prefix --pkg-config-flags=--static --pkg-config=pkg-config --cross-prefix=x86_64-w64-mingw32- --arch=x86_64 --target-os=mingw32 --enable-nonfree --enable-gpl --enable-version3 --disable-debug --disable-w32threads --enable-pthreads --enable-iconv --enable-libxml2 --enable-zlib --enable-libfreetype --enable-libfribidi --enable-gmp --enable-lzma --enable-fontconfig --enable-libvorbis --enable-opencl --disable-libpulse --enable-libvmaf --disable-libxcb --disable-xlib --enable-amf --enable-libaom --enable-avisynth --enable-libdav1d --enable-libdavs2 --enable-libfdk-aac --enable-ffnvcodec --enable-cuda-llvm --enable-frei0r --enable-libgme --enable-libass --enable-libbluray --enable-libmp3lame --enable-libopus --enable-librist --enable-libtheora --enable-libvpx --enable-libwebp --enable-lv2 --enable-libmfx --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenh264 --enable-libopenjpeg --enable-libopenmpt --enable-librav1e --enable-librubberband --enable-schannel --enable-sdl2 --enable-libsoxr --enable-libsrt --enable-libsvtav1 --enable-libtwolame --enable-libuavs3d --disable-libdrm --disable-vaapi --enable-libvidstab --enable-vulkan --enable-libshaderc --enable-libplacebo --enable-libx264 --enable-libx265 --enable-libxavs2 --enable-libxvid --enable-libzimg --enable-libzvbi --extra-cflags=-DLIBTWOLAME_STATIC --extra-cxxflags=--extra-ldflags=-pthread --extra-ldexeflags=--extra-libs=-lgomp --extra-version=20220206  libavutil      57. 19.100 / 57. 19.100  libavcodec     59. 20.100 / 59. 20.100  libavformat    59. 17.101 / 59. 17.101  libavdevice    59.  5.100 / 59.  5.100  libavfilter     8. 26.101 /  8. 26.101  libswscale      6.  5.100 /  6.  5.100  libswresample   4.  4.100 /  4.  4.100  libpostproc    56.  4.100 / 56.  4.100[tcp @ 00000158cdc5bb80] Starting connection attempt to 49.44.85.46 port 443[tcp @ 00000158cdc5bb80] Successfully connected to 49.44.85.46 port 443
[download]  36.9% of 3.91MiB at 113.48KiB/s ETA 00:00bitrate= 116.2kbits/s
total_size=1310720
out_time_us=90201000
out_time_ms=90201000
out_time=00:01:30.201000
dup_frames=0
drop_frames=0
speed= 180x
progress=continue
Input #0, matroska,webm, from 'https://rr3---sn-gwpa-jv3z.googlevideo.com/videoplayback?expire=1649458283&ei=C2hQYq_VMOWI9fwPo8usoAs&ip=49.37.227.151&id=o-AOGmcyslqPHAg9393oMjWNZxeR4XDOekMJGjAVMWY9H_&itag=251&source=youtube&requiressl=yes&mh=qo&mm=31%2C29&mn=sn-gwpa-jv3z%2Csn-gwpa-jwce&ms=au%2Crdu&mv=m&mvi=3&pl=20&gcr=in&initcwndbps=200000&vprv=1&mime=audio%2Fwebm&gir=yes&clen=4099816&dur=243.561&lmt=1540285801235469&mt=1649436303&fvip=3&keepalive=yes&fexp=24001373%2C24007246&c=ANDROID&txp=5511222&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cgcr%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgeRE670-S1-5VxFGq8in-0wKah3ykZrqSNTdwMCvTdqwCIAY7-Fcz1dSWAkt_RWzTLRux_SfbOG1Qk0x8kS28z9uX&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIgWdDRfENdvxBOgMCoUlDYyAKYypOlemnVcVsmaQZQmhgCIQC4mszPG_ibErJr2eqMaC2wAb3IDfwGbf6yqpC33Esxzg%3D%3D':  Metadata:    encoder         : google/video-file  Duration: 00:04:03.56, start: -0.007000, bitrate: 134 kb/s  Stream #0:0(eng): Audio: opus, 48000 Hz, stereo, fltp, delay 312 (default)Output #0, webm, to 'file:Afrojack, Spree Wilson - The Spark (Official Music Video) ft. Spree Wilson [IB3lcPjvWLA].webm.part':  Metadata:    encoder         : Lavf59.17.101  Stream #0:0(eng): Audio: opus, 48000 Hz, stereo, fltp, delay 312 (default)Stream mapping:  Stream #0:0 -> #0:0 (copy)Press [q] to stop, [?] for helpsize=0kB time=00:00:00.00 bitrate=N/A speed=0x
[download]  61.5% of 3.91MiB at 122.56KiB/s ETA 00:00bitrate= 125.5kbits/s
total_size=2359296
out_time_us=150441000
out_time_ms=150441000
out_time=00:02:30.441000
dup_frames=0
drop_frames=0
speed= 149x
progress=continue
size=1280kB time=00:01:30.20 bitrate=116.2kbits/s speed=180x
[download]  76.2% of 3.91MiB at 131.54KiB/s ETA 00:00bitrate= 134.7kbits/s
total_size=3145728
out_time_us=186761000
out_time_ms=186761000
out_time=00:03:06.761000
dup_frames=0
drop_frames=0
speed= 122x
progress=continue
size=2304kB time=00:02:30.44 bitrate=125.5kbits/s speed=149x
[download]  94.3% of 3.91MiB at 124.61KiB/s ETA 00:00bitrate= 127.6kbits/s
total_size=3670016
out_time_us=230181000
out_time_ms=230181000
out_time=00:03:50.181000
dup_frames=0
drop_frames=0
speed= 113x
progress=continue
size=3072kB time=00:03:06.76 bitrate=134.7kbits/s speed=122x
[download] 100% of 3.91MiB in 00:02No more output streams to write to, finishing.
size=    4005kB time=00:04:03.54 bitrate= 134.7kbits/s speed= 107x
video:0kB audio:3920kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 2.172838%
Input file #0 (https://rr3---sn-gwpa-jv3z.googlevideo.com/videoplayback?expire=1649458283&ei=C2hQYq_VMOWI9fwPo8usoAs&ip=49.37.227.151&id=o-AOGmcyslqPHAg9393oMjWNZxeR4XDOekMJGjAVMWY9H_&itag=251&source=youtube&requiressl=yes&mh=qo&mm=31%2C29&mn=sn-gwpa-jv3z%2Csn-gwpa-jwce&ms=au%2Crdu&mv=m&mvi=3&pl=20&gcr=in&initcwndbps=200000&vprv=1&mime=audio%2Fwebm&gir=yes&clen=4099816&dur=243.561&lmt=1540285801235469&mt=1649436303&fvip=3&keepalive=yes&fexp=24001373%2C24007246&c=ANDROID&txp=5511222&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cgcr%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRAIgeRE670-S1-5VxFGq8in-0wKah3ykZrqSNTdwMCvTdqwCIAY7-Fcz1dSWAkt_RWzTLRux_SfbOG1Qk0x8kS28z9uX&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIgWdDRfENdvxBOgMCoUlDYyAKYypOlemnVcVsmaQZQmhgCIQC4mszPG_ibErJr2eqMaC2wAb3IDfwGbf6yqpC33Esxzg%3D%3D):
  Input stream #0:0 (audio): 12178 packets read (4013644 bytes);
  Total: 12178 packets (4013644 bytes) demuxed
Output file #0 (file:Afrojack, Spree Wilson - The Spark (Official Music Video) ft. Spree Wilson [IB3lcPjvWLA].webm.part):
  Output stream #0:0 (audio): 12178 packets muxed (4013644 bytes);
  Total: 12178 packets (4013644 bytes) muxed
[AVIOContext @ 0000029967c0f000] Statistics: 4101289 bytes written, 2 seeks, 17 writeouts
[AVIOContext @ 000002996651e480] Statistics: 4099816 bytes read, 0 seeks
[ffmpeg] Downloaded 4100854 bytes
[download] 100% of 3.91MiB in 00:02
[debug] ffmpeg command line: ffprobe -show_streams "file:Afrojack, Spree Wilson - The Spark (Official Music Video) ft. Spree Wilson [IB3lcPjvWLA].webm"
[ExtractAudio] Destination: Afrojack, Spree Wilson - The Spark (Official Music Video) ft. Spree Wilson [IB3lcPjvWLA].opus
[debug] ffmpeg command line: ffmpeg -y -loglevel "repeat+info" -i "file:Afrojack, Spree Wilson - The Spark (Official Music Video) ft. Spree Wilson [IB3lcPjvWLA].webm" id title formats thumbnails thumbnail description upload_date uploader uploader_id uploader_url channel_id channel_url duration view_count average_rating age_limit webpage_url categories tags playable_in_embed is_live was_live live_status release_timestamp chapters like_count channel channel_follower_count track alt_title availability original_url webpage_url_basename webpage_url_domain extractor extractor_key playlist playlist_index display_id fulltitle duration_string requested_subtitles __has_drm asr filesize format_id format_note source_preference fps height quality tbr url width language language_preference preference ext vcodec acodec dynamic_range abr downloader_options container protocol audio_ext video_ext format resolution http_headers epoch _filename __postprocessors __real_download __finaldir filepath __files_to_move -movflags "+faststart" "file:Afrojack, Spree Wilson - The Spark (Official Music Video) ft. Spree Wilson [IB3lcPjvWLA].opus"
ffmpeg version N-105465-gbedb4bac1c-20220206 Copyright (c) 2000-2022 the FFmpeg developers
  built with gcc 11.2.0 (crosstool-NG 1.24.0.498_5075e1f)
  configuration: --prefix=/ffbuild/prefix --pkg-config-flags=--static --pkg-config=pkg-config --cross-prefix=x86_64-w64-mingw32- --arch=x86_64 --target-os=mingw32 --enable-nonfree --enable-gpl --enable-version3 --disable-debug --disable-w32threads --enable-pthreads --enable-iconv --enable-libxml2 --enable-zlib --enable-libfreetype --enable-libfribidi --enable-gmp --enable-lzma --enable-fontconfig --enable-libvorbis --enable-opencl --disable-libpulse --enable-libvmaf --disable-libxcb --disable-xlib --enable-amf --enable-libaom --enable-avisynth --enable-libdav1d --enable-libdavs2 --enable-libfdk-aac --enable-ffnvcodec --enable-cuda-llvm --enable-frei0r --enable-libgme --enable-libass --enable-libbluray --enable-libmp3lame --enable-libopus --enable-librist --enable-libtheora --enable-libvpx --enable-libwebp --enable-lv2 --enable-libmfx --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenh264 --enable-libopenjpeg --enable-libopenmpt --enable-librav1e --enable-librubberband --enable-schannel --enable-sdl2 --enable-libsoxr --enable-libsrt --enable-libsvtav1 --enable-libtwolame --enable-libuavs3d --disable-libdrm --disable-vaapi --enable-libvidstab --enable-vulkan --enable-libshaderc --enable-libplacebo --enable-libx264 --enable-libx265 --enable-libxavs2 --enable-libxvid --enable-libzimg --enable-libzvbi --extra-cflags=-DLIBTWOLAME_STATIC --extra-cxxflags= --extra-ldflags=-pthread --extra-ldexeflags= --extra-libs=-lgomp --extra-version=20220206
  libavutil      57. 19.100 / 57. 19.100
  libavcodec     59. 20.100 / 59. 20.100
  libavformat    59. 17.101 / 59. 17.101
  libavdevice    59.  5.100 / 59.  5.100
  libavfilter     8. 26.101 /  8. 26.101
  libswscale      6.  5.100 /  6.  5.100
  libswresample   4.  4.100 /  4.  4.100
  libpostproc    56.  4.100 / 56.  4.100
Input #0, matroska,webm, from 'file:Afrojack, Spree Wilson - The Spark (Official Music Video) ft. Spree Wilson [IB3lcPjvWLA].webm':
  Metadata:
    ENCODER         : Lavf59.17.101
  Duration: 00:04:03.56, start: -0.007000, bitrate: 134 kb/s
  Stream #0:0(eng): Audio: opus, 48000 Hz, stereo, fltp (default)
    Metadata:
      DURATION        : 00:04:03.561000000
[NULL @ 000002930b8ed540] Unable to find a suitable output format for 'id'
id: Invalid argument
ERROR: Postprocessing: error running ffmpeg
Traceback (most recent call last):
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\postprocessor\ffmpeg.py", line 527, in run
    self.run_ffmpeg(path, temp_path, acodec, more_opts, information)
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\postprocessor\ffmpeg.py", line 453, in run_ffmpeg
    FFmpegPostProcessor.run_ffmpeg(self, path, out_path, informations, opts)
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\postprocessor\ffmpeg.py", line 343, in run_ffmpeg
    return self.run_ffmpeg_multiple_files([path], out_path, opts, **kwargs)
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\postprocessor\ffmpeg.py", line 300, in run_ffmpeg_multiple_files
    return self.real_run_ffmpeg([(path, []) for path in input_paths], [(out_path, opts)], **kwargs)
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\postprocessor\ffmpeg.py", line 334, in real_run_ffmpeg
    stderr = stderr.strip()
AttributeError: 'NoneType' object has no attribute 'strip'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\YoutubeDL.py", line 3194, in process_info
    replace_info_dict(self.post_process(dl_filename, info_dict, files_to_move))
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\YoutubeDL.py", line 3362, in post_process
    info = self.run_all_pps('post_process', info, additional_pps=info.get('__postprocessors'))
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\YoutubeDL.py", line 3349, in run_all_pps
    info = self.run_pp(pp, info)
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\YoutubeDL.py", line 3322, in run_pp
    files_to_delete, infodict = pp.run(infodict)
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\postprocessor\common.py", line 22, in run
    ret = func(self, info, *args, **kwargs)
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\postprocessor\common.py", line 119, in wrapper
    return func(self, info)
  File "D:\Programs\Source\yt-dlp\yt-dlp\yt_dlp\postprocessor\ffmpeg.py", line 532, in run
    raise PostProcessingError('error running ' + self.basename)
yt_dlp.utils.PostProcessingError: error running ffmpeg
```

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1169,3 +1178,226 @@ def run(self, info):
+        self._duration_to_track, self._total_duration = self._compute_duration_to_track()
+        self._total_filesize = self._compute_total_filesize(self._duration_to_track, self._total_duration)
+        self._status = {
+            'filename': self._info_dict.get('_filename'),
```
1. The file being modified need not the video file. This should be inferred from args
1. For postprocessor hook, the API is not currently defined. To account for multiple files being modified, 'filenames': [...] may be more appropriate
1. Same issue exists for filesize and duration, but worse. Eg: Convert subtitles, convert thumbnails, split chapters, concat video

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1169,3 +1178,226 @@ def run(self, info):
+        start_time, end_time = 0, duration
+        for i, arg in enumerate(self._ffmpeg_args):
+            arg_timestamp = re.match(r'(?P<at>(-ss|-sseof|-to))', arg)
+            if arg_timestamp and i + 1 < len(self._ffmpeg_args):
```
The args can be like `-ss=10`

pukkandan commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -1169,3 +1178,226 @@ def run(self, info):
+            'outputted': 0
+        }
+
+        self.ffmpeg_proc = Popen(
```
there should be a hook here

pukkandan reviewed (changes requested):
Please avoid force-pushing

bashonly commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -836,7 +847,7 @@ def run(self, info):
                 args.extend(['-map', f'{i}:v:0'])
         self.to_screen(f'Merging formats into "{filename}"')
-        self.run_ffmpeg_multiple_files(info['__files_to_merge'], temp_filename, args)
+        self.run_ffmpeg_multiple_files(info['__files_to_merge'], temp_filename, args,  info_dict=info)
```
oops
```suggestion
        self.run_ffmpeg_multiple_files(info['__files_to_merge'], temp_filename, args, info_dict=info)
```

bashonly commented on yt_dlp/postprocessor/ffmpeg.py:
```diff
@@ -202,7 +206,7 @@ def _probe_version(self):
     @property
     def available(self):
-        return self.basename is not None
+        return self._ffmpeg_location.get() or self.basename is not None
```
patching something like this should be done in its own PR
```suggestion
        return self.basename is not None
```

wylieglover commented:
Please @pukkandan @Grub4K, if y'all could review this PR, that would be amazing. Would be great to have the download logger for my project.

Plompi commented:
What are the problems that cause the pull request not to be ready to merge and is there perhaps a way I could help? it would be amazing if this PR would be eventually merged.

arabcoders commented:
I Would like to help work on this as it's really major part of my workflow, if there is anything left to implement i am whiling to try to do it.

Kenshin9977 committed:
Fix FFmpegProgressTracker bugs and regressions

- Restore env= and stdin= params (proxy/piped streams)
- Move Popen out of __init__ into start()
- Init _status in __init__ (prevent AttributeError)
- Remove finished hook from tracker (metaclass handles it)
- Pass output_filename explicitly (not ffmpeg_args[-3])
- Fix multi-stream regex (multiple stream_\d+_\d+_q lines)
- Add -t duration handling in _compute_duration_to_track
- Anchor regexes with fullmatch (duration + time parsing)
- Check smu_parse before hms_parsed (fix dead code path)
- Validate speed string before slicing in _compute_eta
- Gate time.sleep(.5) behind win32 platform check
- Reduce poll interval from 10ms to 50ms
- Remove duplicate debug log
- Replace _streams[bool] with _stdout_data/_stderr_data
- Fix E501 line length violation

Kenshin9977 committed:
Add tests for FFmpegProgressTracker

75 unit tests covering:
- _progress_pattern regex (single/multi-stream, audio-only, N/A)
- ffmpeg_time_string_to_seconds (HMS, ms, us, edge cases)
- _compute_duration_to_track (-ss, -to, -t, -sseof, =form)
- _compute_bitrate (k/m/gbits, N/A, float)
- _compute_eta (normal, N/A speed, zero division)
- _compute_total_filesize (with/without filesize, zero duration)
- Tracker init, status dict, trigger_progress_hook
- _save_stream stdout/stderr separation
- _handle_lines queue draining
- _parse_ffmpeg_output status updates

Kenshin9977 committed:
Fix progress display bugs and harden edge cases

Kenshin9977 committed:
Fix queue race, ETA semantics, console title percent

Kenshin9977 committed:
Fix missing newline in stderr line accumulation

Kenshin9977 committed:
Fix ruff lint errors

Kenshin9977 committed:
Refactor tests: deduplicate helpers, precise assertions, new cases

Kenshin9977 committed:
Fix filesize None crash in FFmpeg merger progress

github-actions[bot] committed:
Merge remote-tracking branch 'upstream/master'

github-actions[bot] committed:
Merge remote-tracking branch 'upstream/master'

github-actions[bot] committed:
Merge remote-tracking branch 'upstream/master'

github-actions[bot] committed:
Merge remote-tracking branch 'upstream/master'

github-actions[bot] committed:
Merge remote-tracking branch 'upstream/master'

github-actions[bot] committed:
Merge remote-tracking branch 'upstream/master'

Kenshin9977 committed:
Merge branch 'master' into master
